kbncontent.go

```diff
@@ -0,0 +1,319 @@
+
+		return "Aggs-based"
+	}
+
+	return "Unknown"
+}
+
+// whether the visualization is considered legacy
+// legacy visualizations should not be used and will be
+// removed from Kibana in the future
+func (v VisualizationDescriptor) IsLegacy() bool {
+	if v.SoType != "visualization" {
+		return false
+	}
+
+	if v.isTSVB() {
+		// TSVB markdown is not marked as legacy because
+		// we don't yet have a good replacement
+		return v.TSVBType() != "markdown"
+	}
+
+	switch v.Type() {
+	case "markdown", "vega":
+		return false
+	default:
+		return true
+	}
+}
+
+func (v VisualizationDescriptor) isTSVB() bool {
+	return v.Type() == "metrics"
+}
+
+// meant to be a visualization-editor-agnostic name for what
+// kind of visualization this actually is (pie, bar, etc)
+// Note: does not yet support Lens
+func (v VisualizationDescriptor) SemanticType() string {
+	if v.isTSVB() {
+		return v.TSVBType()
+	} else {
+		return v.Type()
+	}
+}
+
+// TSVB visualizations are always type "metrics"
+// this property gives the TSVB sub type (gauge, markdown, etc)
+func (v VisualizationDescriptor) TSVBType() string {
+	if !v.isTSVB() {
+		return ""
+	}
+
+	return v.tryDocumentPaths([]string{
+		"attributes.visState.params.type",
+		"embeddableConfig.savedVis.params.type", // by-value dashboard panel
+	})
+}
+
+func (v VisualizationDescriptor) Title() string {
+	if v.SoType != "visualization" {
+		return ""
+	}
+
+	return v.tryDocumentPaths([]string{
+		"attributes.title",
+		"title",
+		"embeddableConfig.savedVis.title", // by-value dashboard panel
+	})
+}
+
+func deserializeSubPaths(doc objx.Map) error {
+	jsonPaths := []string{
+		"attributes.uiStateJSON",
+		"attributes.visState",
+		"attributes.kibanaSavedObjectMeta.searchSourceJSON",
+	}
+	for _, fieldName := range jsonPaths {
+		field := doc.Get(fieldName)
+		if !field.IsStr() {
+			continue
+		}
+		parsed, err := objx.FromJSON(field.Str())
+		if err != nil {
+			return fmt.Errorf("failed to decode embedded json in %q: %w", fieldName, err)
+		}
+		doc.Set(fieldName, parsed)
+	}
+
+	/* these transformations from the original script facilitate the vis_tsvb_aggs and other TSVB-related runtime fields
+		TODO - implement these or convert the
+		vis_tsvb_aggs and any other necessary runtime fields to Go
+	if (
+	    doc?.attributes?.visState?.params?.filter &&
+	    typeof doc.attributes.visState.params.filter !== "string"
+	  ) {
+		  console.log("ENCOUNTERED STRINGIFIED visState.params.filter STATE", path)
+	    doc.attributes.visState.params.filter = JSON.stringify(
+	      doc.attributes.visState.params.filter
+	    );
+	  }
+	  if (
+	    doc?.attributes?.visState?.params?.series &&
+	    Array.isArray(doc.attributes.visState.params.series)
+	  ) {
+		  console.log("ENCOUNTERED STRINGIFIED visState.params.series STATE", path)
+	    doc.attributes.visState.params.series =
+	      doc.attributes.visState.params.series.map((s) => ({
+	        ...s,
+	        filter: JSON.stringify(s.filter),
+	      }));
+	  }
+	*/
```

Comment:
Leftover? Or left here for documentation purposes? If the latter, perhaps add a link to the original code instead.